File: python/starter/src/utils/similarity.py
# calculates the cosine distance between two vectors
import math


def calculate_cosine_similarity(first, second):
    dot_product = sum([a * b for a, b in zip(first, second)])
    magnitude_first = math.sqrt(sum([a**2 for a in first]))
    magnitude_second = math.sqrt(sum([a**2 for a in second]))

    return dot_product / (magnitude_first * magnitude_second)

File: typescript/examples/src/utils/similarity.ts
// computer the cosine similarity between two vectors
export function cosineSimilarity(a: number[], b: number[]): number {
  const dotProduct = a.reduce((acc, val, i) => acc + val * b[i], 0);

  const magnitudeA = Math.sqrt(a.reduce((acc, val) => acc + val * val, 0));
  const magnitudeB = Math.sqrt(b.reduce((acc, val) => acc + val * val, 0));

  return dotProduct / (magnitudeA * magnitudeB);
}
